### PR TITLE
Feautre tips view

### PR DESCRIPTION
**Pull request by AkaShark**
closed https://github.com/tisfeng/Easydict/issues/404

### CODE REVIEW

**Review by Jerry23011 — COMMENTED** [review]
I found that the query service went missing after clearing text after a translation query
截屏2024-02-21 23 51 26

**Comment by AkaShark** [comment]
> I found that the query service went missing after clearing text after a translation query 截屏2024-02-21 23 51 26

Let Me Look at it later

File: Easydict/Feature/ViewController/Cell/EZTableTipsCell.m [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    }];
+    
+    [self.tipsNameLabel mas_makeConstraints:^(MASConstraintMaker *make) {
+        make.left.mas_equalTo(self.tipsIconImageView.mas_right).offset(6);
+        make.centerY.mas_equalTo(self.tipsIconImageView.mas_centerY).offset(1);
+    }];
+    
+//    [self.tipsContentLabel mas_makeConstraints:^(MASConstraintMaker *make) {
+//        make.left.mas_equalTo(9);
+//        make.width.mas_lessThanOrEqualTo(self.bounds.size.width - 9);
```
Comment (phlpsong) [inline Easydict/Feature/ViewController/Cell/EZTableTipsCell.m:66]: Please remove these lines if not needed anymore.

**Comment by phlpsong** [comment]
Do we need to place a close button to let users remove the tips cell manually?

**Review by phlpsong — CHANGES_REQUESTED** [review]

**Comment by AkaShark** [comment]
> Do we need to place a close button to let users remove the tips cell manually?

We don't need to add this close button, when the user query text isn't empty this tips view will not appear. In one query action, it will appear once

**Comment by Jerry23011** [comment]
The previous issue was resolved.

I found that the frame looks weird after starting to type in the query frame (please see screen recording below)

https://github.com/tisfeng/Easydict/assets/89069957/00f01be0-d8ed-4f39-87d1-592a81abacc8

**Comment by tisfeng** [comment]
@AkaShark  Currently, in what cases does this tips view appear?

**Comment by AkaShark** [comment]
> @AkaShark Currently, in what cases does this tips view appear?

When the query window is opened without any text in it, the tips view will appear. The tips view will disappear once the user enters their query text. After the tips view disappears, it will not appear again even if the user clears their query text. It means that the tips view will only appear once per query action.

**Comment by tisfeng** [comment]
I haven't looked at the code carefully yet, for PRs, I usually prefer to run the code first to see the effect. 

Now, I still haven't triggered the display of the tips view, I don't know how to trigger it. I'm confused.

**Comment by AkaShark** [comment]
> I haven't looked at the code carefully yet, for PRs, I usually prefer to run the code first to see the effect.
> 
> Now, I still haven't triggered the display of the tips view, I don't know how to trigger it. I'm confused.

Open EasyDict any window and query text is empty tips view will show, like this
CleanShot 2024-02-29 at 14 05 52@2x

File: Easydict/Feature/ViewController/Cell/EZTableTipsCell.m [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    
+    [super updateConstraints];
+}
+
+- (void)updateQuestionContent {
+    NSArray *questions = self.dataDict[@"questions"];
+    int index = arc4random() % questions.count;
+    self.tipsContentLabel.stringValue = questions[index];
+    NSArray *solves;
+    if ([[EZLanguageManager shared].ezCurrentLanguage isEqualToString:@"zh-Hans"]) {
```
Comment (tisfeng) [inline Easydict/Feature/ViewController/Cell/EZTableTipsCell.m:93]: Do not use the system language to compare directly, the value sometimes ends with an area code, such as `zh-Hans-CN`.

image

Comment (tisfeng) [reply Easydict/Feature/ViewController/Cell/EZTableTipsCell.m:93]: You can use `systemFirstLanguage` instead.

```objc
@property (nonatomic, copy) EZLanguage systemFirstLanguage;
```

Comment (tisfeng) [reply Easydict/Feature/ViewController/Cell/EZTableTipsCell.m:93]: Just use `isSystemChineseFirstLanguage` to check if user system language is Chinese.

**Review by tisfeng — CHANGES_REQUESTED** [review]
When displaying tips view, cell rows are wrong when typing.

https://github.com/tisfeng/Easydict/assets/25194972/1be8397c-e36c-4e45-879b-85c3589a60b7

**Comment by tisfeng** [comment]
We should clear tips view when clicking clear button.

image

**Comment by tisfeng** [comment]
I think we should add an option to allow users to turn off the display of the tips view.

Sometimes the tips view can be annoying if the user is fully aware of the FAQ.

**Comment by AkaShark** [comment]
okay I will perfect the above review on the weekend

**Comment by AkaShark** [comment]
> I think we should add an option to allow users to turn off the display of the tips view.
> 
> Sometimes the tips view can be annoying if the user is fully aware of the FAQ.

I add a toggle  for  appear tips view in advance tab, so user can close tips view by manual.
CleanShot 2024-03-05 at 21 52 28@2x

**Review by tisfeng — CHANGES_REQUESTED** [review]
Looks bad in English mode.

image

**Comment by tisfeng** [comment]
If it's not appropriate to increase the width here, we could shorten the title spelling, `Custom OpenAl Translate` --> `Custom OpenAl`, which I think users should be able to understand.

File: Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m [hunk 3]
```diff
@@ -367,7 +367,7 @@ - (EZQueryService *)defaultTTSService {
 }
 
 - (BOOL)isShowTipsView {
-    if (self.queryModel.actionType != EZActionTypeInputQuery && EZ_isEmptyString(self.queryModel.queryText) && !self.hasShowTips) {
```
Comment (tisfeng) [inline Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m]: Why this removed `self.queryModel.actionType != EZActionTypeInputQuery`?

**Review by tisfeng — CHANGES_REQUESTED** [review]

File: Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m (hunk 3 above)
Comment (AkaShark) [reply Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m]: Is it better to display the tips view when user use input translate? User may get hints from different places

Comment (AkaShark) [reply Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m]: If user don't like the tips view they can close it in advance tab

Comment (tisfeng) [reply Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m]: I fixed this https://github.com/tisfeng/Easydict/pull/418#issuecomment-1971438308 by adding a condition`self.queryModel.actionType != EZActionTypeInputQuery`

File: Easydict/App/Localizable.xcstrings [hunk 4]
```diff
@@ -958,6 +958,23 @@
         }
       }
     },
+    "enable_tips_view" : {
```
Comment (tisfeng) [inline Easydict/App/Localizable.xcstrings:961]: How about changing it to `disable_tips_view`?

File: Easydict/App/Localizable.xcstrings [hunk 5]
```diff
@@ -958,6 +958,23 @@
...
+        "en" : {
+          "stringUnit" : {
+            "state" : "translated",
+            "value" : "Enable Tips View"
+          }
+        },
+        "zh-Hans" : {
+          "stringUnit" : {
+            "state" : "translated",
+            "value" : "开启提示窗口"
```
Comment (tisfeng) [inline Easydict/App/Localizable.xcstrings:973]: `开启提示窗口` --> `关闭提示视图`

File: Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m (hunk 3 above)
Comment (AkaShark) [reply Easydict/Feature/ViewController/Window/BaseQueryWindow/EZBaseQueryViewController.m]: okay, I know it. let me revert it

File: Easydict/App/Localizable.xcstrings (hunk 5 above)
Comment (AkaShark) [reply Easydict/App/Localizable.xcstrings:973]: 😂 you are right

**Review by tisfeng — DISMISSED** [review]

**Review by phlpsong — APPROVED** [review]

**Review by tisfeng — APPROVED** [review]

**Review by phlpsong — APPROVED** [review]

**Comment by tisfeng** [comment]
最近频繁使用，这个功能误触发太多，我先 revert 了，稍后修复好再添加。

**Pull request merged**